sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala

```diff
@@ -500,6 +499,19 @@ case class WholeStageCodegenRDD(@transient sc: SparkContext, var source: CodeAnd
 
   override def compute(split: Partition,
       context: TaskContext): Iterator[InternalRow] = {
+    try {
+      computeInternal(split, context)
+    } catch {
+      case ex: ClassCastException => {
+        // Any other exception in future might be added here
+        CodeGenerator.refresh(source)
+        computeInternal(split, context)
```

Comment:
If we get a ClasssCastException in any other case, that means it's a bug in the plan and it won't be corrected even if we recompile the plan. And we will get the CCE even after the recompile. This retry will only safeguard those problems which arise due to the different classloader.  
Or, did I misunderstood your question ?